Improve `nn.Sequential` documentation

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53380 Improve `nn.Sequential` documentation**

This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

Commit:
Improve ``nn.Sequential`` documentation

[ghstack-poisoned]

SplitInfinity commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
+    the constructor. Alternatively, an ``OrderedDict`` of modules can 
+    be passed in.
+
+    Why should you use a ``Sequential`` instead of a ``ModuleList``? 
+    The value a ``Sequential`` provides over manually calling a sequence
+    of modules is that it allows treating the whole container as a 
+    single module, such that performing a transformation on the 
+    ``Sequential`` applies to each of the modules it stores (which are 
+    each a registered submodule of the ``Sequential``). Also, unlike
+    ``Sequential``,  a ``ModuleList`` does not have a ``forward()`` 
+    method, because it does not define a neural network; in fact, there 
+    is no connection between each of the Modules that it stores. A
+    ``ModuleList`` is exactly what it sounds like--a list for storing
+    ``Module``s! On the other hand, the layers in a ``Sequential`` are 
+    connected in a cascading way.
```
This part is useful!
```suggestion
    Why should you use ``Sequential`` instead of ``ModuleList``? 
    The value ``Sequential`` provides over manually calling a sequence
    of modules is that it allows treating the whole container as a 
    single module, such that performing a transformation on the 
    ``Sequential`` applies to each of the modules it stores (which are 
    each a registered submodule of the ``Sequential``). Also, unlike
    ``Sequential``,  a ``ModuleList`` does not have a ``forward()`` 
    method, because it does not define a neural network; in fact, there 
    is no connection between each of the Modules that it stores. A
    ``ModuleList`` is exactly what it sounds like--a list for storing
    ``Module``s!
```

What do you mean by "transformation"? Also, I prefer `Modules` over `Module`s but I'm not a docs person.

jbschlosser replied:
FYI "`Module`s" is consistent with other docs (requires a space after the backticks and before the "s")

SplitInfinity commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
+    ``Module``s! On the other hand, the layers in a ``Sequential`` are 
+    connected in a cascading way.
+
+    If you're using ``Sequential`` with TorchScript, the inputs of some
+    of the ``Sequential`` submodules may be falsely inferred to be
+    ``Tensor``, even if they're annotated otherwise. This is a known
+    issue in the TorchScript type inference algorithm. The canonical
+    solution is to subclass ``nn.Sequential`` and redeclare ``forward`` 
+    with the input typed correctly.
```
I am of the opinion that this belongs in some TorchScript docs.

jbschlosser replied:
Seconded :)

ansley replied:
Agree, but any ideas as to where I should put it? The best I could find was the [Frequently Asked Questions section](https://pytorch.org/docs/stable/jit.html#frequently-asked-questions) in the TorchScript docs, but it seems like too much of a special case for us to bill it as "frequently asked"

SplitInfinity commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
-    r"""A sequential container.
-    Modules will be added to it in the order they are passed in the constructor.
-    Alternatively, an ordered dict of modules can also be passed in.
+    r"""A list of ``Module``s that acts as a ``Module`` itself.
+    A ``Sequential`` is fundamentally a list of ``Module``s, each with a 
+    ``forward()`` method. ``Sequential`` provides a ``forward()`` method
+    of its own, which accepts any input and forwards it to the first 
+    module it stores. It then "chains" outputs to inputs sequentially 
+    for each subsequent module, finally returning the output of the last
+    module. Modules will be added to it in the order they are passed in 
+    the constructor. Alternatively, an ``OrderedDict`` of modules can 
+    be passed in.
```
```suggestion
    r"""A sequential container.
    Modules will be added to it in the order they are passed in the constructor.
    Alternatively, an ordered dict of modules can also be passed in.
    The ``forward()`` method of ``Sequential`` accepts any input and 
    forwards it to the first module it contains. It then "chains" outputs to inputs 
    sequentially for each subsequent module, finally returning the output of the last
    module.
```

SplitInfinity reviewed (commented):
I don't work on frontend nor docs so I will refrain from approving or requesting changes.

jbschlosser commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
+    the constructor. Alternatively, an ``OrderedDict`` of modules can 
+    be passed in.
+
+    Why should you use a ``Sequential`` instead of a ``ModuleList``? 
+    The value a ``Sequential`` provides over manually calling a sequence
+    of modules is that it allows treating the whole container as a 
+    single module, such that performing a transformation on the 
+    ``Sequential`` applies to each of the modules it stores (which are 
+    each a registered submodule of the ``Sequential``). Also, unlike
+    ``Sequential``,  a ``ModuleList`` does not have a ``forward()`` 
+    method, because it does not define a neural network; in fact, there 
+    is no connection between each of the Modules that it stores. A
+    ``ModuleList`` is exactly what it sounds like--a list for storing
+    ``Module``s! On the other hand, the layers in a ``Sequential`` are 
+    connected in a cascading way.
```
While I agree this section on `Sequential` vs. `ModuleList` has value, I think it's a bit too long, as it now dominates the `Sequential` docs.

I'd prefer if this was trimmed down to 1-2 lines succinctly differentiating `Sequential` and `ModuleList` at a high level, with a link to `ModuleList` so users can read more there if they're interested.

Link can be created with the syntax:
```
:class:`torch.nn.ModuleList`
```

jbschlosser commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
-    r"""A sequential container.
-    Modules will be added to it in the order they are passed in the constructor.
-    Alternatively, an ordered dict of modules can also be passed in.
+    r"""A list of ``Module``s that acts as a ``Module`` itself.
+    A ``Sequential`` is fundamentally a list of ``Module``s, each with a 
```
There are some formatting errors here, which can be seen in the generated doc artifacts (FYI these can be found by clicking "Details" next to the `pytorch_python_doc_build` check and selecting the Artifacts tab): https://11343754-65600975-gh.circle-artifacts.com/0/docs/generated/torch.nn.Sequential.html

I think for "`Module`s" to properly render, there should be a space after the backticks and before the "s".

jbschlosser commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
+    Example::
 
-        # Example of using Sequential
+        # Using Sequential to create a small model. When `model` is run,
+        # input will first be passed to `Conv2d(1,20,5)`, then to 
+        # `ReLU`, then to `Conv2d(20,64,5)`, then to `ReLU` again
```
This wording is slightly confusing to me; it sounds like the same initial input will be passed to each item in the list. I think the description should better match the chaining behavior you described well in the first paragraph.

Also, tiny nit: "then to `ReLU` again" -> "then to the second `ReLU`" or something to be unambiguous about the linear chaining behavior.

jbschlosser commented on torch/nn/modules/container.py:
```diff
@@ -27,21 +27,50 @@ def __init__(self, **kwargs: Any) -> None:
 
-        # Example of using Sequential with OrderedDict
+        # Using Sequential with OrderedDict. This is functionally the
+        # same as the above code
```
Seems like ending periods are left out from other example comments (e.g. [`Conv2d` docs](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html))

jbschlosser reviewed (commented):
Thanks for beefing these docs up! :)

nit: Please make sure the trailing whitespace is removed

Commit:
Update on "Improve `nn.Sequential` documentation"


This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

[ghstack-poisoned]

Commit:
Update on "Improve `nn.Sequential` documentation"


This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

[ghstack-poisoned]

Commit:
Update on "Improve `nn.Sequential` documentation"


This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

[ghstack-poisoned]

Commit:
Update on "Improve `nn.Sequential` documentation"


This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

[ghstack-poisoned]

SplitInfinity commented on docs/source/jit.rst:
```diff
@@ -629,6 +629,16 @@ Q: I would like to trace module's method but I keep getting this error:
 
     To trace a specific method on a module, see :func:`torch.jit.trace_module <torch.jit.trace_module>`
 
+Gotchas
```
The meaning of this term might be confusing to some readers. Consider revising.

SplitInfinity reviewed (approved):
Signing off on the JIT part.

jbschlosser reviewed (approved):
Module-side changes LGTM!

Commit:
Update on "Improve `nn.Sequential` documentation"


This PR was primarily a response to [[JIT] nn.Sequential of nn.Module with input type List[torch.Tensor] inferred to torch.Tensor](https://github.com/pytorch/pytorch/issues/52588), but I added a bit more information about ``nn.Sequential`` as well

Differential Revision: [D26849861](https://our.internmc.facebook.com/intern/diff/D26849861)

[ghstack-poisoned]
